libr/bin/format/elf/elf.c

```diff
@@ -1277,228 +1299,189 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
+	ut64 addr_jmprel = bin->dyn_info.dt_jmprel;
+	ut64 addr_got = bin->dyn_info.dt_mips_pltgot;
+
+	if (!addr_jmprel || !addr_got) {
+		return UT64_MAX;
 	}
 
-	const int htsize = R_MIN (nrel, 1024);
-	rel_cache = ht_up_new_size (htsize, NULL, rel_cache_free, NULL);
+	ut64 pos = (rel->rva - addr_got - 0x2 * WORDSIZE) / WORDSIZE;
 
-	for (j = k = 0; j < rel_sec->size && k < nrel; j += tsize, k++) {
-		int len = r_buf_read_at (bin->b, rel_sec->offset + j, rl, tsize);
-		if (len != tsize) {
-			break;
-		}
-		struct ht_rel_t *rel = read_ht_rel (bin, rl, k);
-		if (!rel) {
-			goto out;
-		}
+	ut8 buf[1024];
+	const ut8 *base;
+	ut64 plt_addr = addr_jmprel + bin->dyn_info.dt_pltrelsz;
+	ut64 p_plt_addr = Elf_(r_bin_elf_v2p_new) (bin, plt_addr);
+	int res = r_buf_read_at (bin->b, p_plt_addr, buf, sizeof (buf));
+	if (res != sizeof (buf)) {
+		return UT64_MAX;
+	}
 
-		if (!ht_up_insert (rel_cache, REL_SYM, rel)) {
-			free (rel);
-		}
+	base = r_mem_mem_aligned (buf, sizeof (buf), (const ut8 *)"\x3c\x0f\x00", 3, 4);
+	if (base) {
+		plt_addr += (int)(size_t) (base - buf);
+	} else {
+		plt_addr += MIPS_PLT_OFFSET + 8; // HARDCODED HACK
 	}
-	return rel_cache;
-out:
-	ht_up_free (rel_cache);
-	return NULL;
+	plt_addr += pos * 16;
+
+	return plt_addr;
 }
 
-static ut64 get_dyn_entry(ELFOBJ *bin, int dyn_entry) {
-	int i;
-	for (i = 0; i < bin->dyn_entries; i++) {
-		if (bin->dyn_buf[i].d_tag == dyn_entry) {
-			switch (bin->dyn_buf[i].d_tag) {
-			case DT_REL:
-			case DT_RELA:
-			case DT_PLTGOT:
-			case DT_JMPREL:
-				return bin->dyn_buf[i].d_un.d_ptr;
-			case DT_RELSZ:
-			case DT_RELASZ:
-			case DT_PLTRELSZ:
-				return bin->dyn_buf[i].d_un.d_val;
-			default:
-				r_warn_if_reached ();
-				break;
-			}
-		}
-	}
+static size_t get_size_rel_mode(size_t rel_mode) {
+	return rel_mode == DT_RELA? sizeof (Elf_(Rela)): sizeof (Elf_(Rel));
+}
 
-	return -1;
+static size_t get_num_relocs_dynamic_plt(ELFOBJ *bin) {
+	if (bin->dyn_info.dt_pltrelsz) {
+		return bin->dyn_info.dt_pltrelsz / get_size_rel_mode (bin->dyn_info.dt_pltrel);
+	}
+	return 0;
 }
 
-static ut64 get_got_addr(ELFOBJ *bin) {
-	ut64 ret = get_dyn_entry (bin, DT_PLTGOT);
-	if (ret != -1) {
-		return ret;
+static ut64 get_import_addr_riscv(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 got_addr = bin->dyn_info.dt_pltgot;
+	if (!got_addr) {
+		return UT64_MAX;
 	}
 
-	ret = Elf_(r_bin_elf_get_section_addr) (bin, ".got");
-	if (ret) {
-		return ret;
+	ut64 plt_addr = get_got_entry (bin, rel);
+	if (plt_addr == UT64_MAX) {
+		return UT64_MAX;
 	}
 
-	ret = Elf_(r_bin_elf_get_section_addr) (bin, ".got.plt");
-	if (ret) {
-		return ret;
+	ut64 pos = (rel->rva - got_addr - 0x2 * WORDSIZE) / WORDSIZE;
+	return plt_addr + RISCV_PLT_OFFSET + pos * 0x10;
+}
+
+static ut64 get_import_addr_sparc(ELFOBJ *bin, RBinElfReloc *rel) {
+	if (rel->type != R_SPARC_JMP_SLOT) {
+		bprintf ("Unknown sparc reloc type %d\n", rel->type);
+		return UT64_MAX;
+	}
+	ut64 tmp = get_got_entry (bin, rel);
+	if (tmp == UT64_MAX) {
```

Comment:
just use a single return statemnt like this: `return (tmp==UT64_MAX)? tmp-6:UT64_MAX